### ISSUE TITLE
Program hangs when triggering page fault

### ISSUE THREAD

**Issue by Marsman1996** [asterinas/asterinas#1535]
### Describe the bug
Program hangs in Asterinas when triggering page fault, and exits (or crashes) in Linux.

### To Reproduce
1. Compile a program which trigger page fault:
```C
#include <setjmp.h>
#include <signal.h>
#include <stdio.h>

jmp_buf env;

void handle_sigsegv(int sig) {
  printf("Segmentation fault occurred!\n");
  longjmp(env, 1);
}

int main() {
  signal(SIGSEGV, handle_sigsegv);

  int fault_count = 0;
  const int max_faults = 10;

  while (fault_count < max_faults) {
    if (setjmp(env) == 0) {
      int *invalid_ptr = (int *)0xDEADBEEF;
      printf("Will trigger a segmentation fault in %x\n", invalid_ptr);
      *invalid_ptr = 42;
    } else {
      fault_count++;
      printf("Recovered from segmentation fault #%d\n", fault_count);
    }
  }

  printf("Completed after triggering %d segmentation faults.\n", fault_count);
  return 0;
}
```
2. Run the program, it hangs in Asterinas, and exits (or crashes) in Linux.

### Expected behavior
The program shall not hang in Asterinas.

### Environment
- Official docker asterinas/asterinas:0.9.4
- Intel(R) Xeon(R) Gold 6230R CPU @ 2.10GHz
- Asterinas version: main 7e2e9ce

### Logs
In debug mode, it repeated with
```
[32m[  1917.212][0m [93mWARN [0m: [39mpage fault handler failed: addr: 0xdeadbeef, err: Error { errno: EACCES, msg: Some("page fault addr is not in current vmar") }[0m
```

**Comment by Marsman1996** [comment]
Asterinas fall into infinite loop in `user_task_entry()`:
https://github.com/asterinas/asterinas/blob/efd49a96e3eb83094f6bc93f2723ef376b29fabb/kernel/src/thread/task.rs#L64-L88

Because Asterinas fail to handle page fault in `handle_page_fault_from_vmar()`:
https://github.com/asterinas/asterinas/blob/efd49a96e3eb83094f6bc93f2723ef376b29fabb/kernel/src/thread/exception.rs#L60-L69

As a result it generate a fault signal `aster_nix::process::signal::signals::fault::FaultSignal {num: aster_nix::process::signal::sig_num::SigNum {sig_num: 11}, code: 1, addr: core::option::Option<u64>::Some(3735928559)}` and enqueue it.
https://github.com/asterinas/asterinas/blob/efd49a96e3eb83094f6bc93f2723ef376b29fabb/kernel/src/thread/exception.rs#L33-L39

**Comment by Marsman1996** [comment]
@boterinas label "help wanted"

**Comment by boterinas** [comment]
**Error**: Parsing relabel command in [comment](https://github.com/asterinas/asterinas/issues/1535#issuecomment-2469820256) failed: ...' label' | error: a label delta at >| ' "help wan'...

Please file an issue on GitHub at [triagebot](https://github.com/rust-lang/triagebot) if there's a problem with this bot, or reach out on [#t-infra](https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra) on Zulip.

**Comment by Marsman1996** [comment]
@boterinas label help wanted

**Comment by boterinas** [comment]
**Error**: Label help can only be set by Rust team members

Please file an issue on GitHub at [triagebot](https://github.com/rust-lang/triagebot) if there's a problem with this bot, or reach out on [#t-infra](https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra) on Zulip.

**Comment by Marsman1996** [comment]
A new PoC here, after running this, crtl-C or crtl-Z not works

```C
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/mman.h>
#include <errno.h>

#define SWAP_FILE "./swapfile"
#define SWAP_FILE_SIZE (1024 * 1024 * 128) // 128 MB

int main(void) {
    // Turn off swap for the file if it exists.
    if (swapoff(SWAP_FILE) == -1 && errno != EINVAL) {
        perror("swapoff failed");
        // return 1;
    }

    // Remove the swap file if it exists.
    if (unlink(SWAP_FILE) == -1 && errno != ENOENT) {
        perror("unlink failed");
        return 1;
    }

    // Create and zero-fill the swap file.
    int fd = open(SWAP_FILE, O_CREAT | O_WRONLY | O_CLOEXEC, 0600);
    if (fd == -1) {
        perror("swap file open failed");
        return 1;
    }

    // Use fallocate to allocate and zero-fill the file.
    if (fallocate(fd, 0, 0, SWAP_FILE_SIZE) == -1) {
        perror("fallocate failed");
        close(fd);
        return 1;
    }

    close(fd);

    // Set up the swap file.
    char cmdline[64];
    sprintf(cmdline, "mkswap %s", SWAP_FILE);
    if (system(cmdline) == -1) {
        perror("mkswap failed");
        return 1;
    }

    // Enable the swap file.
    if (swapon(SWAP_FILE, 0) == -1) {
        perror("swapon failed");
        return 1;
    }

    printf("Swap file set up and enabled successfully.\n");
    return 0;
}

```

**Comment by cchanging** [comment]
For case 1, I think it may be a corner case of signal handler recursion. If changing the handler to the following:
```c
void handle_sigsegv(int sig) {
  printf("Segmentation fault occurred!\n");
  int *invalid_ptr = (int *)0xDEADBEEF;
  printf("trigger a segmentation fault recursion in %x\n", invalid_ptr);
}
``` 
The program will also crash in Linux and hang in Asterinas. This is because we allow the nested user signal handler. Once the thread get a SIGSEGV, it will enter the handler and get a SIGSEGV again. 

Linux also allow the nested user signal handler, if we change the handler to 
```c
void handle_sigsegv(int sig) {
  printf("Segmentation fault occurred!\n");
  raise(SIGSEGV);
}
```
It will also lead the program to hang in Linux. I think there may be a special treatment in Linux for the "signal handler - exception - signal handler" situation, but I don't know what it is yet...

**Comment by Marsman1996** [comment]
Case 2 (mkswap) seems to be fixed.

**Comment by Marsman1996** [comment]
> Linux also allow the nested user signal handler, if we change the handler to
> 
> void handle_sigsegv(int sig) {
>   printf("Segmentation fault occurred!\n");
>   raise(SIGSEGV);
> }
> It will also lead the program to hang in Linux. I think there may be a special treatment in Linux for the "signal handler - exception - signal handler" situation, but I don't know what it is yet...

However, in your case, the `SIGSEGV` could be caught by `handle_sigsegv` 🤔

**Comment by cchanging** [comment]
> However, in your case, the SIGSEGV could be caught by handle_sigsegv 🤔

The original handler, which triggers a invalid memory access, will also generate SIGSEGV. It will be caught by `handle_sigsegv` again in Asterinas, but in Linux it will lead to a crash. I think this may be the different.

**Comment by Marsman1996** [comment]
> > However, in your case, the SIGSEGV could be caught by handle_sigsegv 🤔
> 
> The original handler, which triggers a invalid memory access, will also generate SIGSEGV. It will be caught by `handle_sigsegv` again in Asterinas, but in Linux it will lead to a crash. I think this may be the different.

But if it's caught again by `handle_sigsegv`, shouldn’t we see the endless printf output, similar to what happens with `raise(SIGSEGV);`?

**Comment by cchanging** [comment]
> But if it's caught again by handle_sigsegv, shouldn’t we see the endless printf output, similar to what happens with raise(SIGSEGV);?

You're right, I made a mistake before. During the execution of a signal handler, that particular signal is masked. In Asterinas' case, since SEGV is masked while in its handler, it cannot be re-triggered and the execution simply returns to the faulting instruction, causing an endless page fault loop. 

I suspect Linux has some mechanism, perhaps directly crashing when page faults (or similar exceptions) occur within a SEGV handler. I'll check the kernel source code to verify.
